### PR TITLE
Allow forcing compression with v2s2 format

### PR DESCRIPTION
**Pull request by Jamstah**
This highlights a bug in common where the compression format is reset if the format is v2s2, even if its a valid compression format.

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: test/e2e/push_test.go [hunk 1]
```diff
@@ -141,6 +141,44 @@ var _ = Describe("Podman push", func() {
...
+		push := podmanTest.Podman([]string{"push", "-q", "--tls-verify=false", "--force-compression=true", "--compression-format", "gzip", "--format", "v2s2", "--remove-signatures", "imageone", "localhost:5005/image"})
+		push.WaitWithDefaultTimeout()
+		Expect(push).Should(ExitCleanly())
+
+		skopeoInspect := []string{"inspect", "--tls-verify=false", "--raw", "docker://localhost:5005/image:latest"}
+		skopeo := SystemExec("skopeo", skopeoInspect)
+		skopeo.WaitWithDefaultTimeout()
+		Expect(skopeo).Should(ExitCleanly())
+		output := skopeo.OutputToString()
+		Expect(output).ToNot(ContainSubstring("zstd"))
```
Comment (Luap99) [inline test/e2e/push_test.go:175]: This should check for the proper gzip mime type I would assume? Checking for not zstd does not mean it cannot be any other incorrect value.

File: test/e2e/push_test.go [hunk 2]
```diff
@@ -141,6 +141,44 @@ var _ = Describe("Podman push", func() {
...
+
+		push = podmanTest.Podman([]string{"push", "-q", "--tls-verify=false", "--force-compression=true", "--compression-format", "zstd", "--format", "v2s2", "--remove-signatures", "imageone", "localhost:5005/image"})
+		push.WaitWithDefaultTimeout()
+		Expect(push).Should(ExitWithError(125, "cannot use ForceCompressionFormat with undefined default compression format"))
```
Comment (Luap99) [inline test/e2e/push_test.go:179]: This error seems extremely unhelpful. Why `undefined default compression format`? The command clearly specified zstd as format so this makes just no sense.

**Review by Luap99 — COMMENTED** [review]
cc @mtrmac

File: test/e2e/push_test.go (hunk 2 above)
Comment (Jamstah) [reply test/e2e/push_test.go:179]: Yes, that's just testing that the change to common doesn't mean that other invalid cases will still be attempted.

The fix to common is only to allow it to work in the gzip case, not to fix the misleading error completely.

File: test/e2e/push_test.go [hunk 3]
```diff
@@ -141,6 +141,44 @@ var _ = Describe("Podman push", func() {
...
+		defer lock.Unlock()
+		session := podmanTest.Podman([]string{"run", "-d", "--name", "registry", "-p", "5005:5000", REGISTRY_IMAGE, "/entrypoint.sh", "/etc/docker/registry/config.yml"})
+		session.WaitWithDefaultTimeout()
+		Expect(session).Should(ExitCleanly())
+
+		if !WaitContainerReady(podmanTest, "registry", "listening on", 20, 1) {
+			Skip("Cannot start docker registry.")
+		}
+
+		session = podmanTest.Podman([]string{"build", "-t", "imageone", "build/basicalpine"})
```
Comment (mtrmac) [inline test/e2e/push_test.go:162]: @Luap99 Do we need to (or, do for the future, do we want to) worry about running the tests concurrently? 

AFAICS they would clash in this image tag, as well as the `registry` container name.

**Comment by martinpitt** [comment]
Wrt. failed cockpit rawhide test: I reported the criu/kernel regression to https://bugzilla.redhat.com/show_bug.cgi?id=2328985 and silenced it in our tests in https://github.com/cockpit-project/bots/pull/7146 . From now on, this failure will be ignored, i.e. tests should go back to green. Please either retry or re-push.

File: test/e2e/push_test.go (hunk 2 above)
Comment (mtrmac) [reply test/e2e/push_test.go:179]: Yes, this is all a fallout of the c/common code being a short-term hack (where we didn’t do the long-term correct thing), see https://github.com/containers/common/pull/1869 .

What we _want_ is to differentiate between compression formats coming from `containers.conf` (which should be silently ignored for destinations which can’t support them, silently falling back to `gzip`) and compression formats coming from the command line (which should cause an error when the user asks for an impossible operation).

But the podman CLI (here https://github.com/containers/podman/blob/ceee7cb0a6b92412f6be5d6206cd4ac764cd99e6/cmd/podman/images/push.go#L140 ) is in the habit of using the `containers.conf` values as CLI option defaults, making it, from the very start, hard to tell the difference (although, IIRC, possible). And by the time the data gets to c/common, there is no longer any way to tell, in the current c/common API.

---

I do think that a comment here, in the test case, to the effect of “the command is asking for an impossible thing; per https://github.com/containers/common/pull/1869, we should detect that and fail with a precise error, but that does not exist, so we end up reporting this misleading text” would be useful.

**Review by mtrmac — COMMENTED** [review]
Thanks! ACK to the goal and overall approach.

**Comment by mtrmac** [comment]
For the record, this PR exists to test https://github.com/containers/common/pull/2255 .

File: test/e2e/push_test.go (hunk 3 above)
Comment (Luap99) [reply test/e2e/push_test.go:162]: Everything in test/e2e is already running concurrently (with very few exceptions, look for Serial decorator). The way e2e tests are setup is that each test case has its own storage (--root/--runroot/--tmpdir) so there are never any conflicts with names so this should be fine here.

Comment (Jamstah) [reply test/e2e/push_test.go:162]: Isn't that why the registry for each test is starter on a different port?

**Comment by mtrmac** [comment]
/approve

LGTM

**Comment by mtrmac** [comment]
Thanks @Jamstah !

**Pull request merged**